slides/E-async-web.md

```diff
@@ -0,0 +1,892 @@
+    runtime.block_on(run());
+}
+```
+
+*Question: What stands out to you? What strikes you as odd?*
+
+---
+layout: default
+---
+
+# Async in Rust
+
+- Revolve around `Future` trait (~like JS `Promise`, C# `Task`)  
+  &rarr; `async fn`s return `Future`s
+
+- `Future`s are inert
+- `async` is zero-cost
+- No built-in runtime
+- Single- or multithreaded execution
+- Can be mixed with other concurrency models
+- Relatively new and lacks some features and nice diagnostics
+
+---
+layout: default
+---
+
+# State of the `async` art
+What you can expect doing `async` Rust
+
+- Blazingly fast applications
+- More interaction with advanced language features
+- Compatibility issues (re: colored functions)
+- Faster evolving ecosystem
+- `async fn` in Traits not stable
+
+*But still a work in progress*
+
+---
+layout: default
+---
+
+# Support of `async`
+
+- Fundamental types and traits are in `std`
+- `async`/`await` are native to the language
+- Utilities/extensions in `futures` crate
+- Async runtimes are third party
+
+Example runtimes: `async-std`, `tokio`, `smol`
+
+---
+layout: section
+---
+# The `Future` trait
+Foundation of async
+
+---
+layout: two-cols
+---
+
+# A `VerySimpleFuture`
+```rust
+trait VerySimpleFuture {
+    type Output;
+    /// Do work and check if task is completed.
+    /// Returns [Poll::Ready], containing the 
+    /// `Output` if task is ready,
+    /// [Poll::Pending] if not
+    fn poll(&mut self) -> Poll<Self::Output>;
+}
+
+enum Poll<T> {
+    Ready(T),
+    Pending,
+}
+```
+::right::
+<v-click>
+
+```rust
+struct VerySimpleAlarm {
+    alarm_time: Instant,
+}
+
+impl VerySimpleFuture for VerySimpleAlarm {
+    type Output = ();
+
+    fn poll(&mut self) -> Poll<()> {
+        if Instant::now() >= self.alarm_time {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+```
+</v-click>
+
+---
+layout: two-cols
+---
+
+# Executing `VerySimpleFuture`
+ ```rust
+fn main() {
+    let mut first_alarm = VerySimpleAlarm {
+        alarm_time: Instant::now() 
+            + Duration::from_secs(3)
+    };
+    let mut snooze_alarm = VerySimpleAlarm {
+        alarm_time: Instant::now() 
+            + Duration::from_secs(5)
+    };
+
+    loop {
+        if let Poll::Ready(_) = first_alarm.poll() {
+            println!("Beep beep beep");
+        }
+        if let Poll::Ready(_) = snooze_alarm.poll() {
+            println!("You're late for work!")
+        }
+    }
+}
+ ```
+::right::
+<v-click>
+<div>
+
+```txt
+[pause...]
+Beep beep beep
+Beep beep beep
+[... a few moments later...]
+You're late for work!
+Beep beep beep
+You're late for work!
+Beep beep beep
+You're late for work!
+[...ad infinitum]
+```
+
+## It works! 🎉
+
+*Question: How can `VerySimpleFuture` be improved?*
+</div>
+</v-click>
+
+---
+layout: default
+---
+# Limitation
+
+- Busy waiting
+- How to signal the executor the future is *actually* ready to be polled?
+
+<v-click>
+<br/>
+
+## ⏰ Introduce a Waker
```

Comment:
If you're sleeping past an alarm clock, you're more likely a very deep `Dreamer` though. 😄 

More seriously, with the slide title I was expecting the `Waker` to be introduced, but since that is the runtime part you are obviously not doing that (yet); so I would think about a different title than "Introduce". No concrete suggestion atm.